### PR TITLE
Cpp optimization

### PR DESCRIPTION
**Pull request by hkalodner**
This PR shouldn't change any behavior other than eliminating some really annoying hangs. It appears that when run through CGo, the current AVM implementation uses more stack space than go allows leading to it hanging (presumably due to undefined behavior of some sort). This issue has been plaguing us for a while, but was quite difficult to track down due to appearing quite hard to recreate and isolate.

Moving forward, we should make sure to not introduce and functions into the CPP codebase that use large amount of stack space through excessive iteration.

- Converts many recursive value and tuple methods to iterative forms to avoid stack overflows
- Refactor TuplePool system and turn it into a singleton
- Assorted other optimizations

### CODE REVIEW

**Comment by hkalodner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 3
           

Clones added
============
- packages/arb-avm-cpp/avm_values/src/vmValueParser.cpp  2
- packages/arb-avm-cpp/tests/checkpoint.cpp  2
- packages/arb-avm-cpp/avm_values/include/avm_values/opcodes.hpp  2
         

Clones removed
==============
+ packages/arb-avm-cpp/tests/datastack.cpp  -6
+ packages/arb-avm-cpp/tests/opcodes.cpp  -5
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/OffchainLabs/arbitrum/pullRequest?prid=5984944&bid=19625735)

**Comment by hkalodner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 3
           

Clones added
============
- packages/arb-avm-cpp/tests/checkpoint.cpp  2
- packages/arb-avm-cpp/avm_values/include/avm_values/opcodes.hpp  2
- packages/arb-avm-cpp/avm_values/src/vmValueParser.cpp  2
         

Clones removed
==============
+ packages/arb-avm-cpp/tests/datastack.cpp  -6
+ packages/arb-avm-cpp/tests/opcodes.cpp  -5
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/OffchainLabs/arbitrum/pullRequest?prid=5984944&bid=19625735)

File: packages/arb-avm-cpp/avm_values/src/pool.cpp [hunk 1]
```diff
@@ -44,8 +63,28 @@ std::shared_ptr<RawTuple> TuplePool::getResource(size_t s) {
     return resource;
 }
 
-void TuplePool::returnResource(std::shared_ptr<RawTuple>&& object) {
-    if (!shuttingDown) {
-        resources[object->data.size()].push_back(std::move(object));
+void TuplePool::deleteTuple(UniqueTuple tup) {
+    if (shuttingDown) {
```
Comment (edfelten) [inline packages/arb-avm-cpp/avm_values/src/pool.cpp:67]: Empty if statement shouldn't exist.  Should there be code inside?  Or should it be deleted?

File: packages/arb-avm-cpp/avm_values/src/tuple.cpp [hunk 2]
```diff
@@ -114,47 +105,79 @@ Tuple::Tuple(value val1,
...
     }
 }
 
-void Tuple::marshal(std::vector<unsigned char>& buf) const {
-    buf.push_back(TUPLE + tuple_size());
-    for (uint64_t i = 0; i < tuple_size(); i++) {
-        marshal_value(get_element(i), buf);
+constexpr uint64_t hash_size = 32;
+
+struct BasicValChecker {
```
Comment (edfelten) [inline packages/arb-avm-cpp/avm_values/src/tuple.cpp:119]: Leave a comment explaining what property of the object the "ValChecker" is checking.

File: packages/arb-avm-cpp/avm/include/avm/machinestate/datastack.hpp [hunk 3]
```diff
@@ -26,7 +26,7 @@ class Transaction;
 struct SaveResults;
 
 class Datastack {
-    static constexpr int lazyCount = 1000;
+    static constexpr int lazyCount = 10000;
```
[the remark this answers is not in the document]
Comment (hkalodner) [reply packages/arb-avm-cpp/avm/include/avm/machinestate/datastack.hpp:33]: Added a comment

File: packages/arb-avm-cpp/avm_values/src/pool.cpp (hunk 1 above)
Comment (hkalodner) [reply packages/arb-avm-cpp/avm_values/src/pool.cpp]: I removed the dead code

File: packages/arb-avm-cpp/avm_values/src/tuple.cpp (hunk 2 above)
Comment (hkalodner) [reply packages/arb-avm-cpp/avm_values/src/tuple.cpp:122]: Added a comment

**Comment by hkalodner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 3
           

Clones added
============
- packages/arb-avm-cpp/tests/checkpoint.cpp  2
- packages/arb-avm-cpp/avm_values/include/avm_values/opcodes.hpp  2
- packages/arb-avm-cpp/avm_values/src/vmValueParser.cpp  2
         

Clones removed
==============
+ packages/arb-avm-cpp/tests/opcodes.cpp  -5
+ packages/arb-avm-cpp/tests/datastack.cpp  -6
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/OffchainLabs/arbitrum/pullRequest?prid=5984944&bid=19625735)

**Comment by hkalodner** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 1
- Added 3
           

Clones added
============
- packages/arb-avm-cpp/avm_values/include/avm_values/opcodes.hpp  2
- packages/arb-avm-cpp/tests/checkpoint.cpp  2
- packages/arb-avm-cpp/avm_values/src/vmValueParser.cpp  2
         

Clones removed
==============
+ packages/arb-avm-cpp/tests/datastack.cpp  -6
+ packages/arb-avm-cpp/tests/opcodes.cpp  -5
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/OffchainLabs/arbitrum/pullRequest?prid=5984944&bid=19625735)

**Pull request merged**